src/raftstore/store/store.rs

```diff
@@ -1538,6 +1942,63 @@ impl<T: Transport, C: PdClient> Store<T, C> {
         }
     }
 
+    /// Check if a request is valid if it has premerge/merge proposal.
+    fn check_merge_proposal(&self, msg: &mut RaftCmdRequest) -> Result<()> {
+        if !msg.get_admin_request().has_pre_merge() && !msg.get_admin_request().has_merge() {
+            return Ok(());
+        }
+
+        let region_id = msg.get_header().get_region_id();
+        let peer = &self.region_peers[&region_id];
+        let region = peer.region();
+
+        if msg.get_admin_request().has_pre_merge() {
+            let target_region = msg.get_admin_request().get_pre_merge().get_target();
+            let peer = match self.region_peers.get(&target_region.get_id()) {
+                None => return Err(box_err!("target region doesn't exist.")),
+                Some(p) => p,
+            };
+            if peer.region() != target_region {
+                return Err(box_err!("target region not matched, skip proposing."));
+            }
+            if !util::is_sibling_regions(target_region, region) {
+                return Err(box_err!("regions are not sibling, skip proposing."));
+            }
+            if !util::region_on_same_store(target_region, region) {
+                return Err(box_err!(
+                    "peers doesn't match {:?} != {:?}, reject merge",
+                    region.get_peers(),
+                    target_region.get_peers()
+                ));
+            }
+        } else {
+            let source_region = msg.get_admin_request().get_merge().get_source();
+            let source_peer = &self.region_peers[&source_region.get_id()];
+            // only merging peer can propose merge request.
+            assert!(
+                source_peer.pending_merge.is_some(),
+                "{} {} should be in merging state",
+                peer.tag,
+                source_peer.tag
+            );
+            assert_eq!(source_region, source_peer.region());
+            assert!(
+                util::is_sibling_regions(source_region, region),
```

Comment:
Can't target region split after propose `per_merge` on source region?

Reply:
Yes, it can. In that case, `PreMerge` will be rollback.